craft_parts/plugins/rust_plugin.py

```diff
@@ -85,9 +85,14 @@ def validate_environment(
 
         :param part_dependencies: A list of the parts this part depends on.
         """
-        if "rust-deps" in (part_dependencies or []):
-            options = cast(RustPluginProperties, self._options)
-            if options.rust_channel and options.rust_channel != "none":
+        options = cast(RustPluginProperties, self._options)
+        has_rust_deps = "rust-deps" in (part_dependencies or [])
+        if has_rust_deps or options.rust_channel == "none":
```

Comment:
This logic is a bit confusing to me. Bearing in mind that [`validate_environment()` gets run just before the build step](https://github.com/canonical/craft-parts/blob/394ede95879c3f4d65704efce45d70762581c739/craft_parts/executor/part_handler.py#L460-L470), it seems to me that we can simply check for `cargo` and `rustc` and not bother to check for rustup (since all we're using it for is to ensure we have the correct `cargo` and `rustc` configured).

If `rustup` were improperly configured, the `pull` step should have failed before we run this method.

Reply:
So if the user has specified `rustc` and `cargo` in `rust-deps` part, when `validate_environment` gets called, are these packages supposed to be already installed at this point?

Reply:
Also the logic here is that if `rust-deps` exists or the user specified not to use `rustup`, we will check if `cargo` and `rustc` binaries exist in the environment.

Reply:
I stand corrected on this, as it looks like `validate_environment()` also gets run during the prologue. So the conditional is correct. (Our testing infrastructure should be improved to handle those cases, but that's a separate matter.)

I think if the user wants to provide their own toolchain, they should download it in a `rust-deps` part, so that shouldn't be an issue.

Thanks for your patience!